sdk/core/Azure.Core/tests/RequestContextTests.cs

```diff
@@ -27,5 +44,244 @@ public void CanSetErrorOptions()
+            context.AddPolicy(new AddHeaderPolicy("PerRetryHeader", "PerRetry"), HttpPipelinePosition.PerRetry);
+            context.AddPolicy(new AddHeaderPolicy("BeforeTransportHeader", "BeforeTransport"), HttpPipelinePosition.BeforeTransport);
+
+            // TODO: Should be Get<Resource>
+            Response response = await client.GetAsync("snoopy", context);
+
+            Request request = mockTransport.Requests[0];
+
+            Assert.IsTrue(request.Headers.TryGetValues("PerCallHeader1", out var perCall1Values));
+            Assert.AreEqual(1, perCall1Values.Count());
+            Assert.AreEqual("PerCall1", perCall1Values.ElementAt(0));
+
+            Assert.IsTrue(request.Headers.TryGetValues("PerCallHeader2", out var perCall2Values));
+            Assert.AreEqual(1, perCall2Values.Count());
+            Assert.AreEqual("PerCall2", perCall2Values.ElementAt(0));
+
+            Assert.IsTrue(request.Headers.TryGetValues("PerRetryHeader", out var perRetryValues));
+            Assert.AreEqual("PerRetry", perRetryValues.ElementAt(0));
+            Assert.AreEqual("PerRetry", perRetryValues.ElementAt(1));
+            Assert.AreEqual("PerRetry", perRetryValues.ElementAt(2));
+
+            Assert.IsTrue(request.Headers.TryGetValues("BeforeTransportHeader", out var beforeTransportValues));
+            Assert.AreEqual("BeforeTransport", beforeTransportValues.ElementAt(0));
+            Assert.AreEqual("BeforeTransport", beforeTransportValues.ElementAt(1));
+            Assert.AreEqual("BeforeTransport", beforeTransportValues.ElementAt(2));
+        }
+
+        [Test]
+        public async Task CanAddPolicies_ThreeWays()
+        {
+            var retryResponse = new MockResponse(408); // Request Timeout
+            var resourceResponse = new MockResponse(200);
+            var resource =
+            @"{
+                name = ""snoopy"",
+                id = ""beagle""
+            }";
+            resourceResponse.SetContent(resource);
+
+            var mockTransport = new MockTransport(retryResponse, resourceResponse);
+            var options = new TemplateServiceClientOptions()
+            {
+                Transport = mockTransport,
+            };
+
+            var perCallPolicies = new HttpPipelinePolicy[] { new AddHeaderPolicy("PerCall", "Builder") };
+            var perRetryPolicies = new HttpPipelinePolicy[] { new AddHeaderPolicy("PerRetry", "Builder") };
+
+            options.AddPolicy(new AddHeaderPolicy("BeforeTransport", "ClientOptions"), HttpPipelinePosition.BeforeTransport);
+            options.AddPolicy(new AddHeaderPolicy("PerRetry", "ClientOptions"), HttpPipelinePosition.PerRetry);
+            options.AddPolicy(new AddHeaderPolicy("PerCall", "ClientOptions"), HttpPipelinePosition.PerCall);
+
+            var pipeline = HttpPipelineBuilder.Build(options, perCallPolicies, perRetryPolicies, null);
```

Comment:
Consider https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/tests/HttpPipelineBuilderTest.cs or https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/tests/HttpPipelineTests.cs